### PR TITLE
8345179: RISC-V: Add gtests for weak cmpxchg

### PR DESCRIPTION
**Pull request by robehn**
Hi, please consider.

This adds tests of aligned weak narrow cmpxchg.
```
[ RUN      ] RiscV.cmpxchg_weak_int16_lr_sc_vm
[       OK ] RiscV.cmpxchg_weak_int16_lr_sc_vm (2 ms)
[ RUN      ] RiscV.cmpxchg_weak_int8_lr_sc_vm
[       OK ] RiscV.cmpxchg_weak_int8_lr_sc_vm (0 ms)
[ RUN      ] RiscV.cmpxchg_weak_int16_maybe_zacas_vm
[       OK ] RiscV.cmpxchg_weak_int16_maybe_zacas_vm (0 ms)
[ RUN      ] RiscV.cmpxchg_weak_int8_maybe_zacas_vm
[       OK ] RiscV.cmpxchg_weak_int8_maybe_zacas_vm (0 ms)
[----------] 4 tests from RiscV (20997 ms total)
```
Executed with -XX:+UnlockExperimentalVMOptions -XX:+UseZacas

Thanks, Robbin

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8345179](https://bugs.openjdk.org/browse/JDK-8345179): RISC-V: Add gtests for weak cmpxchg (**Sub-task** - P4)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22476/head:pull/22476` \
`$ git checkout pull/22476`

Update a local copy of the PR: \
`$ git checkout pull/22476` \
`$ git pull https://git.openjdk.org/jdk.git pull/22476/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22476`

View PR using the GUI difftool: \
`$ git pr show -t 22476`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22476.diff">https://git.openjdk.org/jdk/pull/22476.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22476#issuecomment-2511496456)
</details>

### CODE REVIEW

**Comment by RealFYang** [comment]
Hi, Seems this covers `MacroAssembler::weak_cmpxchg_narrow_value`. Are we going to cover this one: `MacroAssembler::cmpxchg_weak`?

**Comment by robehn** [comment]
> Hi, Seems this only covers `MacroAssembler::weak_cmpxchg_narrow_value`. Are we going to add testing for this one: `MacroAssembler::cmpxchg_weak`?

As the Zacas is the same as normal cmpxchg and the lr/sc just changes the loop to failure it wasn't my focus.

I can add it if you want, to this PR or a later.

File: test/hotspot/gtest/riscv/test_assembler_riscv.cpp [hunk 1]
```diff
@@ -260,4 +260,133 @@ TEST_VM(RiscV, cmpxchg_int32_plain_maybe_zacas) {
...
+    return ret;
+  }
+  
+  static TESTSIZE weak_cmpxchg(intptr_t addr, TESTSIZE expected, TESTSIZE new_value) {
+    BufferBlob* bb = BufferBlob::create("riscvTest", 128);
+    CodeBuffer code(bb);
+    MacroAssembler _masm(&code);
+    address entry = _masm.pc();
+    {
+       _masm.cmpxchg_weak(/*addr*/ c_rarg0, /*expected*/ c_rarg1, /*new_value*/ c_rarg2,
```
Comment (RealFYang) [inline test/hotspot/gtest/riscv/test_assembler_riscv.cpp:295]: Seems that we should rename macro-assembler routine `cmpxchg_weak` to `weak_cmpxchg`.
I have prepared another PR to fix this naming issue: https://github.com/openjdk/jdk/pull/22505.

File: test/hotspot/gtest/riscv/test_assembler_riscv.cpp [hunk 2]
```diff
@@ -260,4 +260,133 @@ TEST_VM(RiscV, cmpxchg_int32_plain_maybe_zacas) {
...
+    }
+    _masm.flush(); // icache invalidate
+    TESTSIZE ret = ((cmpxchg_func)entry)(addr, expected, new_value, /*result*/ 67);
+    BufferBlob::free(bb);
+    return ret;
+  }
+};
+
+template <typename TESTSIZE, Assembler::operand_size ASMSIZE>
+void run_narrow_cmpxchg_tests() {
```
[the remark this answers is not in the document]
Comment (robehn) [reply test/hotspot/gtest/riscv/test_assembler_riscv.cpp]: Fixed

File: test/hotspot/gtest/riscv/test_assembler_riscv.cpp (hunk 1 above)
Comment (robehn) [reply test/hotspot/gtest/riscv/test_assembler_riscv.cpp:295]: Fixed

File: test/hotspot/gtest/riscv/test_assembler_riscv.cpp [hunk 3]
```diff
@@ -260,4 +260,133 @@ TEST_VM(RiscV, cmpxchg_int32_plain_maybe_zacas) {
...
+}
+
+TEST_VM(RiscV, cmpxchg_weak_int8_maybe_zacas) {
+  if (UseZacas) {
+    run_narrow_cmpxchg_tests<int8_t, Assembler::int8>();
+  }
+}
+
+template <typename TESTSIZE, Assembler::operand_size ASMSIZE>
+void weak_cmpxchg_test() {
```
[the remark this answers is not in the document]
Comment (robehn) [reply test/hotspot/gtest/riscv/test_assembler_riscv.cpp]: Fixed

File: test/hotspot/gtest/riscv/test_assembler_riscv.cpp (hunk 1 above)
Comment (robehn) [reply test/hotspot/gtest/riscv/test_assembler_riscv.cpp:295]: Ok!

**Review by RealFYang — APPROVED** [review]
Thanks for the update. BTW: You need to fix the trailing whitespace issue reported by jcheck.

File: test/hotspot/gtest/riscv/test_assembler_riscv.cpp [hunk 4]
```diff
@@ -260,4 +260,133 @@ TEST_VM(RiscV, cmpxchg_int32_plain_maybe_zacas) {
...
+    ASSERT_EQ(data[i], 42);
+
+    data[i] = 121;
+    ret = WeakCmpxchgTester<TESTSIZE, ASMSIZE>::weak_narrow_cmpxchg((intptr_t)&data[i], 120, 42);
+    ASSERT_EQ(ret, 0);
+    ASSERT_EQ(data[i], 121);
+  }
+}
+
+TEST_VM(RiscV, cmpxchg_weak_int16_lr_sc) {
```
Comment (Hamlin-Li) [inline test/hotspot/gtest/riscv/test_assembler_riscv.cpp:327]: I'm not familiar with the gtest syntax in JDK, seems this (`cmpxchg_weak_int16_lr_sc`) is just a name?
But it still good to follow the name convenction, i.e. weak_cmpxchg.

File: test/hotspot/gtest/riscv/test_assembler_riscv.cpp [hunk 5]
```diff
@@ -260,4 +260,133 @@ TEST_VM(RiscV, cmpxchg_int32_plain_maybe_zacas) {
...
+  ASSERT_EQ(ret, 1);
+  ASSERT_EQ(data, 42);
+
+  data = 121;
+  ret = WeakCmpxchgTester<TESTSIZE, ASMSIZE>::weak_cmpxchg((intptr_t)&data, 120, 42);
+  ASSERT_EQ(ret, 0);
+  ASSERT_EQ(data, 121);
+}
+
+TEST_VM(RiscV, cmpxchg_weak_int64_lr_sc) {
```
Comment (Hamlin-Li) [inline test/hotspot/gtest/riscv/test_assembler_riscv.cpp:366]: similar comment here, and below

**Review by Hamlin-Li — COMMENTED** [review]
Looks good, thanks for adding the tests.
Just one minor comment.

**Comment by RealFYang** [comment]
Hi, Please merge with latest master as https://github.com/openjdk/jdk/pull/22505 has been integrated.

File: test/hotspot/gtest/riscv/test_assembler_riscv.cpp (hunk 4 above)
Comment (robehn) [reply test/hotspot/gtest/riscv/test_assembler_riscv.cpp]: fixed in merge

File: test/hotspot/gtest/riscv/test_assembler_riscv.cpp (hunk 5 above)
Comment (robehn) [reply test/hotspot/gtest/riscv/test_assembler_riscv.cpp]: fixed in merge

**Review by Hamlin-Li — APPROVED** [review]
Thanks for updating, looks good.

**Review by RealFYang — APPROVED** [review]

**Comment by robehn** [comment]
Thanks!

**Pull request closed without merging**